rust-runtime/aws-smithy-http/src/callback.rs

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! A module for traits that define callbacks that will be called at specific points in an HTTP request's lifecycle.
+
+mod checksum;
+
+use http::{HeaderMap, HeaderValue};
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// A callback that, when inserted into a request body, will be called for corresponding lifecycle events.
+pub trait BodyCallback: Send {
+    /// This lifecycle function is called for each chunk **successfully** read. If an error occurs while reading a chunk,
+    /// this method will not be called. This method takes `&mut self` so that implementors may modify an implementing
+    /// struct/enum's internal state. Implementors may return an error.
+    fn update(&mut self, #[allow(unused_variables)] bytes: &[u8]) -> Result<(), BoxError> {
```

Comment:
just for visual noise, might be nicer to add `let _ = bytes;` in the body instead of the unused variables allow